fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathComputer.java

```diff
@@ -0,0 +1,227 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.rule.tree.prunesubfield;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.starrocks.catalog.ColumnAccessPath;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptExpressionVisitor;
+import com.starrocks.sql.optimizer.base.ColumnRefSet;
+import com.starrocks.sql.optimizer.operator.Operator;
+import com.starrocks.sql.optimizer.operator.OperatorBuilderFactory;
+import com.starrocks.sql.optimizer.operator.logical.LogicalAggregationOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalCTEConsumeOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalJoinOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalProjectOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalScanOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalUnionOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rewrite.ReplaceColumnRefRewriter;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/*
+ * compute all complex column access path in whole plan, we only need check the expression
+ * which one in project and predicate (the rule execute order promise this)
+ */
+class SubfieldAccessPathComputer extends OptExpressionVisitor<OptExpression, Void> {
+    private final Map<ScalarOperator, ColumnRefSet> allComplexColumns = Maps.newHashMap();
+
+    private void visitPredicate(OptExpression optExpression) {
+        if (optExpression.getOp().getPredicate() == null) {
+            return;
+        }
+
+        SubfieldExpressionCollector collector = new SubfieldExpressionCollector();
+        optExpression.getOp().getPredicate().accept(collector, null);
+        for (ScalarOperator expr : collector.getComplexExpressions()) {
+            allComplexColumns.put(expr, expr.getUsedColumns());
+        }
+    }
+
+    private OptExpression visitChildren(OptExpression optExpression, Void context) {
+        for (int i = optExpression.getInputs().size() - 1; i >= 0; i--) {
+            OptExpression child = optExpression.inputAt(i);
+            optExpression.setChild(i, child.getOp().accept(this, child, context));
+        }
+        return optExpression;
+    }
+
+    @Override
+    public OptExpression visit(OptExpression optExpression, Void context) {
+        visitPredicate(optExpression);
+        return visitChildren(optExpression, context);
+    }
+
+    @Override
+    public OptExpression visitLogicalProject(OptExpression optExpression, Void context) {
+        LogicalProjectOperator lpo = optExpression.getOp().cast();
+        rewriteColumnsInPaths(lpo.getColumnRefMap(), true);
```

Comment:
is it also applied to common expression in project?

Reply:
here has no common expression